metadata-ingestion/tests/unit/test_kafka_source.py

```diff
@@ -26,407 +25,463 @@
-
-        ctx = PipelineContext(run_id="test1")
-        kafka_source = KafkaSource.create(
-            {
-                "topic_patterns": {"allow": ["test"]},
-                "connection": {"bootstrap": "localhost:9092"},
-            },
-            ctx,
-        )
-        workunits = [w for w in kafka_source.get_workunits()]
+@patch("datahub.ingestion.source.kafka.confluent_kafka.Consumer", autospec=True)
+def test_kafka_source_configuration(mock_kafka):
+    ctx = PipelineContext(run_id="test")
+    kafka_source = KafkaSource(
+        KafkaSourceConfig.parse_obj({"connection": {"bootstrap": "foobar:9092"}}),
+        ctx,
+    )
+    kafka_source.close()
+    assert mock_kafka.call_count == 1
 
-        mock_kafka.assert_called_once()
-        mock_kafka_instance.list_topics.assert_called_once()
-        assert len(workunits) == 2
 
-        mock_cluster_metadata.topics = ["test", "test2", "bazbaz"]
-        ctx = PipelineContext(run_id="test2")
-        kafka_source = KafkaSource.create(
-            {
-                "topic_patterns": {"allow": ["test.*"]},
-                "connection": {"bootstrap": "localhost:9092"},
-            },
-            ctx,
-        )
-        workunits = [w for w in kafka_source.get_workunits()]
-        assert len(workunits) == 4
-
-    @patch("datahub.ingestion.source.kafka.confluent_kafka.Consumer", autospec=True)
-    def test_kafka_source_workunits_with_platform_instance(self, mock_kafka):
-        PLATFORM_INSTANCE = "kafka_cluster"
-        PLATFORM = "kafka"
-        TOPIC_NAME = "test"
-
-        mock_kafka_instance = mock_kafka.return_value
-        mock_cluster_metadata = MagicMock()
-        mock_cluster_metadata.topics = [TOPIC_NAME]
-        mock_kafka_instance.list_topics.return_value = mock_cluster_metadata
-
-        ctx = PipelineContext(run_id="test1")
-        kafka_source = KafkaSource.create(
-            {
-                "connection": {"bootstrap": "localhost:9092"},
-                "platform_instance": PLATFORM_INSTANCE,
-            },
-            ctx,
-        )
-        workunits = [w for w in kafka_source.get_workunits()]
-
-        # We should only have 1 topic + sub-type wu.
-        assert len(workunits) == 2
-        assert isinstance(workunits[0], MetadataWorkUnit)
-        assert isinstance(workunits[0].metadata, MetadataChangeEvent)
-        proposed_snap = workunits[0].metadata.proposedSnapshot
-        assert proposed_snap.urn == make_dataset_urn_with_platform_instance(
-            platform=PLATFORM,
-            name=TOPIC_NAME,
-            platform_instance=PLATFORM_INSTANCE,
-            env="PROD",
-        )
+@patch("datahub.ingestion.source.kafka.confluent_kafka.Consumer", autospec=True)
+def test_kafka_source_workunits_wildcard_topic(mock_kafka):
+    mock_kafka_instance = mock_kafka.return_value
+    mock_cluster_metadata = MagicMock()
+    mock_cluster_metadata.topics = {"foobar": None, "bazbaz": None}
```

Comment:
list_topics().topics returns a dictionary with topic name as key and TopicMetadata as value.